### ISSUE TITLE
Voxel GI creates glowing shadow edges

### ISSUE THREAD

**Issue by WickedInsignia** [godotengine/godot#83025]
### Godot version

4.2 Dev 5

### System information

Windows 11, Nvidia RTX4070ti, AMD Ryzen7700x

### Issue description

Voxel GI spills light into shadows in an unrealistic manner, creating a "glow" effect around their edges and propagating light across surfaces that should not be affected by the bounces they're creating. This gives meshes an unintentionally waxy appearance, since this effect is similar to subsurface scattering.
It also spills light around corners into areas unaffected by direct light, causing light to leak around sharp angles.

VoxelGI at default settings (128 subdiv). The Glow post processing effect is disabled, so the light leaking into the shadow the structure is casting is caused by Voxel GI:
![VoxelGIGlowingShadows_128](https://github.com/godotengine/godot/assets/70243134/bf1b5212-35f8-45c2-9063-8b1bcd61e609)

This is not resolved by higher resolutions (512). Here we can see the light bleeding around sharp angles more clearly, near the top of the structure:
![VoxelGIGlowingShadows_512](https://github.com/godotengine/godot/assets/70243134/d5a7302b-fc7f-4ea6-9a65-de466e05ea54)

We can see the effect in the Voxel lighting view (128 for this example):
![VoxelGIGlowingShadows_128Voxel](https://github.com/godotengine/godot/assets/70243134/9257b90b-107a-4903-a861-c887b98e4575)

Changing Normal Bias can reduce the effect, but causes further leaks in other areas (128):
![VoxelGIGlowingShadows_128NormalBias](https://github.com/godotengine/godot/assets/70243134/2a8a3966-1f69-4c1b-aed7-d5e438834114)

When we use Blender's Cycles as a ground truth example, there is no light bleeding into the shadows. Only light bouncing onto surrounding surfaces:
![VoxelGIGlowingShadows_BlenderRender](https://github.com/godotengine/godot/assets/70243134/f58eb733-1141-4007-8fd2-139eedd8a3ec)

Thus, the optimal result would be as follows. Light shouldn't bleed into shadowed areas on the surface they are cast against nor should it bleed around corners. It should only bounce onto other surfaces:
![VoxelGIGlowingShadows_128VoxelRevised](https://github.com/godotengine/godot/assets/70243134/17feca91-00dc-4fcf-8bb0-0a2bd1bbba4a)
![VoxelGIGlowingShadows_128VoxelRevised02](https://github.com/godotengine/godot/assets/70243134/53eab1e2-3fbe-4517-97e8-760c9a15d926)

### Steps to reproduce

Simply use VoxelGI with any dynamic light (directional, spot, omni) with shadows turned on.

### Minimal reproduction project

[VoxelGI_GlowingShadows.zip](https://github.com/godotengine/godot/files/12842389/VoxelGI_GlowingShadows.zip)

**Comment by Calinou** [comment]
Given how VoxelGI works with a lot of approximations, this may not be entirely fixable.

Static geometry is baked so it could technically be possible to perform some kind of supersampling (at the cost of longer bake times), but we have few people knowledgeable with VoxelGI.

**Comment by WickedInsignia** [comment]
Yeah totally understandable. If there's any way for it to determine whether a surface/normal is where the direct light is hitting and bouncing _from_, there might be some way to tell it not to spread light further onto that surface and simply bounce it away.
I expect preventing light bleeding around corners to be more tricky since you're working with a single voxel in that area, there's no way to create a sharp "corner." Either there's a line of light voxels in the shadow area, or a line of dark voxels in the light area.

Regardless I'm seeing it "radiate" light further around corners than it should. I don't see any reason (as an artist, not as a developer who deeply understands VoxelGI) why the slightly lighter voxels at the top edge of this building can't be as dark as the ones directly below them:
![BistroVoxel02](https://github.com/godotengine/godot/assets/70243134/2204c7d0-1aa5-4c73-9749-0e5952627f6c)
![BistroVoxel01](https://github.com/godotengine/godot/assets/70243134/915f09c0-43db-4bb7-88ce-facca4ad67c8)

This also seems to be a situation of dark voxels inside the area that's being hit with light (the entire roof of that building) without negative effects.
Again just an artist so I wouldn't be too surprised if I completely misunderstand what the limitations of improving this are.
